### DISCUSSION TITLE
`next/image` feature request: generate specific sizes for each image

### DISCUSSION THREAD

**Discussion by vtenfys** [vercel/next.js#32140] [Ideas]
### Describe the feature you'd like to request

Rather than providing `deviceSizes` and `imageSizes` in `next.config.js`, I would like these (i.e. the `srcset` attribute) to be generated automatically based on the `sizes` prop of `<Image>`.

Currently, I am using `next/image` in the following manner:

```js
// inside a <div> which handles positioning at different breakpoints
<Image
  layout="responsive"
  sizes="(min-width: 1024px) 413px, (min-width: 640px) 256px, 384px"
  src={coverImage.url}
  alt={coverImage.altText}
  width={2000}
  height={2000}
/>
```

I'm using the `sizes` prop to specify the maximum image size at each breakpoint, based on my responsive design.

And this is my `next.config.js`:

```js
module.exports = {
  images: {
    domains: ['media.graphcms.com'],
    deviceSizes: [], // specify all possible image sizes manually
    imageSizes: [256, 384, 413].flatMap((px) => [px, px * 2, px * 3]),
  },
}
```

I am setting `deviceSizes` to an empty array since I am not using any full-width images, therefore additional sizes would be unnecessary.

I am generating three sizes for each CSS pixel width, to account for different device pixel ratios (DPRs).

### Describe the solution you'd like

Rather than having to specify `imageSizes` manually in `next.config.js`, the `srcset` attribute of `img` should be populated automatically based on the `sizes` of a given image.

This reduces duplication, and avoids having an ever-growing list of sizes which are applied to every single `<Image>` despite that some sizes may only be needed for a particular UI element.

I would suggest instead being able to configure the supported DPRs, with a default of `[1, 2, 3]`. This corresponds with the `flatMap()` in my above implementation.

#### Caveats

Since this approach may not be suitable when `layout` is set to `intrinsic` or `fixed`, I suggest that `imageSizes` should still be available to configure, but ignored when `layout` is set to `responsive` or `fill`.

### Describe alternatives you've considered

My current implementation is the primary alternative to my proposed solution, which is flawed because it involves adding more sizes to every image across the whole site, for each new responsive layout (unless the image sizes are the same).

Another alternative would be to allow manually specifying the `srcset` for each individual `<Image>`, however this feels cumbersome given that the `<Image>` component is supposed to simplify the creation of responsive images.

**Comment by vtenfys** [comment]
For now, I've implemented my own custom component to achieve my desired outcome:

```js
/**
 * @param {Object} props
 * @param {string} props.sizes
 * @param {string} props.src
 * @param {string} props.alt
 * @param {number} props.width
 * @param {number} props.height
 */
export default function Image({ sizes, src, alt, width, height, ...props }) {
  const quality = 90
  const widths = sizes
    .split(', ')
    .map((size) => parseInt(size.match(/(?:\(.+\) )?([0-9]+)px/)[1]))
    .flatMap((size) => [size, size * 2, size * 3]) // different DPRs
  const maxWidth = Math.max(...widths)
  const srcSet = widths.map((w) => getSrc(src, w, quality) + ` ${w}w`)

  return (
    // eslint-disable-next-line @next/next/no-img-element
    <img
      {...props}
      loading="lazy"
      sizes={sizes}
      src={getSrc(src, maxWidth, quality)}
      srcSet={srcSet.join(', ')}
      alt={alt}
      width={width}
      height={height}
    />
  )
}

/**
 * @param {string} src
 * @param {number} width
 * @param {number} quality
 */
function getSrc(src, width, quality) {
  const url = encodeURIComponent(src)
  return `/_next/image?url=${url}&w=${width}&q=${quality}`
}
```

This still requires setting all possible sizes in `next.config.js` due to the image service rejecting requests containing unspecified widths, however it avoids every single `<img>` containing all these widths regardless of usage.

**Comment by psiho** [comment]
I'm surprised this thread is dead for so long. Just starting a first project with Next and amount of unnecessary items ins `srcset` bothers me. Are we over optimizing here? Individual images are no big deal, but long lists with thumbnails? Hmm.
